### DISCUSSION TITLE
I2C Slave does not work (STM32F103)

### DISCUSSION THREAD

**Discussion by AndKe** [stm32duino/Arduino_Core_STM32#2608] [Q&A]
**Describe the bug**
I2C slave does not recieve requests.

**To Reproduce**

```c++
#include <Wire.h>
long lastms;
volatile int x = 0;
volatile int y = 0;

static void receiveEvent(int howMany) {
  x = Wire.read();
}

static void requestEvent() {
  y = 1;
}

void setup() {
  Serial.begin(115200);

  pinMode(PB6, INPUT); //PB6=SCL
  pinMode(PB7, INPUT); //PB7=setSDA
  pinMode(PC13, OUTPUT);

  Wire.setSCL(PB6);
  Wire.setSDA(PB7);
  Wire.begin(11);                // join i2c bus as slave
  //Wire.onRequest(requestEvent);  // register event
  Wire.onReceive(receiveEvent);  // register event
  lastms = millis();
}

void loop() {
  if (millis() > lastms + 1000) {
    Serial.print("-");
    lastms = millis();

  }
  if (x != 0) {
    Serial.print("U:0x");
    Serial.println(x, HEX);
    //digitalWrite(PC13,!digitalRead(PC13));
    x = 0;
  }
  if (y != 0) {
    Serial.println("V");
    y = 0;
  }
}

```

1. Run the code om STM32F103Cx
2. the code expects requests from master to address x0xB / 11d
3. I see the data requests using a oscolloscope/logic analyzer on PB6/PB7
4. The software outputs only "-"

**Expected behavior**
I'd expect it to work like it does on Atmega/atmel controllers - where this code works.   see https://github.com/arduino/ArduinoCore-API/issues/241

Than I rebuilt it all using Arduino Core 2.9.0  (current release)
On Core 2.9.0 - the PB6 is at least not pulled low - which is good, but it still does not work.

![image](https://github.com/user-attachments/assets/e8343448-63da-4e78-bfe6-81a187731424)

**Desktop (please complete the following information):**
 - OS: Linux Ubuntu24.10
 - Arduino IDE version: 2.3.3
 - STM32 core version: 2.9.0
 - Tools menu settings if not the default: [e.g. Newlib Standard, No Serial]
 - Upload method: SWD

**Board (please complete the following information):**
 - Name: Bluepill STM32F103C8
 - Hardware Revision: n/a
 - Extra hardware used if any: -

**Comment by fpistm** [comment]
Do you have pullup resistors on each I2c lines?

**Comment by AndKe** [comment]
Yes, and - I can see the I2C packets using an oscilloscope on the pins. (decoded on the scope)  
Plus : for simple troubleshooting I hooked it up on a breadboard. When I do the very same setup, but switch the blackpill an Atmega based arduino boards - it works.

**Comment by fpistm** [comment]
Could you share the master example?

**Comment by AndKe** [comment]
@fpistm unable, the master is Ardupilot (which is open source)- if you have some hardware that can run Ardupilot, and configure BATT_MONITOR 5  - then you have that I2C traffic too.

The traffic is really nothing special - it is all about spotting a normal 7-bit address that is being interrogated by a master.  I bet this is reproducible with a ny simple Arduino program that tried to read data from a I2C device (just set the address to the same on both ends.)

**Comment by AndKe** [comment]
Please see attached communication capture with device 0xB    - the capture should clarify any questions you might have about protocol/timing etc. it is captured using Saleae logic analyzer, which is perfectly multi-platform with nice software: https://www.saleae.com/pages/downloads

[BMSEmu Boot.zip](https://github.com/user-attachments/files/17973614/BMSEmu.Boot.zip)

**Comment by fpistm** [comment]
Thanks for inputs. I will try to test on my side but don't know when as I'm fully busy on other tasks.

**Comment by AndKe** [comment]
Thank you, I am really stuck on this one, I "modernized" previous design (Atmega32u4) by dropping in STM32F103, and now - with plenty of new PCB's manufactured, I am stuck with this issue.
I have previously had projects with STM32F103C8 and CB  acting as master without any issues.

**Comment by AndKe** [comment]
Additional information; the exactly same happens when SCL=PB10 and SDA=PB11

**Comment by AndKe** [comment]
@fpistm I understand that you are busy, but is there some way to ...motivate... you or another competent contributor here to solve this?

**Comment by fpistm** [comment]
Just a question you said you used a BluePill but you select a blackpill on your screenshot ?

**Comment by AndKe** [comment]
Yes, the blackpill is an error on my side.
The device in question is something that did not work as I2C slave after being ported from Atmega32U4  And all issues here are about STM32F103C8T6.  be it on my hardware, or Bluepill.

After much troubleshooting I made this test app above, and ran it on Atmega32u4 for comparsion/to verify it.
Then I also moved from my board to an actual bluepill (just to easily test on PB10/PB11) .
During the compilations for various targets, I must have messed it up.
I will retest that once I get home, in about 2 hours.

BTW: I have also tried to build it on VSCode w/platformio (and Arduino library)  - with same results. So me trying both Arduino and all that are kind of desperate attempts to make it work :)

**Comment by AndKe** [comment]
@fpistm corrected, and I can confirm that setting "Bluepill" as target did not change the result.

**Comment by ItsXor** [comment]
I can confirm this issue exists.
Only input I can give is that Wire Slave sometimes accepts data right after reset, maybe first 300ms.
This issue is not present in other core (https://github.com/rogerclarkmelbourne/Arduino_STM32).
Would love this to be resolved because SimpleFOC only works with this core, but I can't control my Bluepill as I2C slave simultaneously.

**Reply by fpistm** [reply]
@ItsXor 
They are no issue, I guess it is simply a wrong usage...
Duplicate: https://community.st.com/t5/others-hardware-and-software/bug-in-arduino-core-stm32-i2c-slave/m-p/752220
Opening a topic on ST community while discussion already opened here is time consuming for me as all will end on my huge todo  list...

**Comment by AndKe** [comment]
@ItsXor Thank you! - for making me doubt my sanity.  I started to wonder how this was possible, without anyone spotting it before.
I would love to know what can be brought over from that core into PlatformIO  (as my project is really there, and I do not really know if PlatformIO can somehow use that core without much hacking.
On the other hand, I just hope one of the skilled STM engineers can trow together a fix soon.

**Comment by AndKe** [comment]
@fpistm  Dear STM - please note: "all I want for Christmas is a fix for I2C Slave"   - this is the losest "writing to Santa" i ever did.  :)

**Comment by fpistm** [comment]
Will try to work on this soon anyway any help will be appreciated.

**Comment by AndKe** [comment]
@fpistm very glad to hear that. Unfortunately I won't be of much help nailing the issue(insufficient skills in the area), but I definitely can help testing or running a debug build.

**Comment by fpistm** [comment]
@AndKe 
I've start to check your issue.
What is exactly your issue? 
Your talk about:
> 2- the code expects requests from master to address x0xB / 11d

but `onRequest` is comment on your code and see no `Wire.write` on your code?
So I don't understand your issue.


I've tested all the  examples from Wire for Master (send/receive) and Slave(send/receive) and they work.

**Comment by fpistm** [comment]
Here test sketch:

<details>
<summary>Master</summary>

```c++
#include <Wire.h>

#define I2C_ADDR  11

static byte x = 1;

void setup()
{
  Wire.begin();        // join i2c bus (address optional for master)
  Serial.begin(9600);  // start serial for output
}

void loop()
{
  Wire.requestFrom(I2C_ADDR, 1);  // request 1 bytes from slave device

  while(Wire.available())         // slave may send less than requested
  {
    char c = Wire.read();         // receive a byte as character
    Serial.println(c, HEX);              // print the character
  }

  delay(10);

  Wire.beginTransmission(I2C_ADDR); // transmit to device
  Wire.write(x);                    // sends one byte
  Wire.endTransmission();           // stop transmitting
  x++;

  delay(500);
}
```
</details>

<details>
<summary>Slave based on your sketch</summary>

```c++
#include <Wire.h>
long lastms;
volatile int x = 0;
volatile int y = 0;

static void receiveEvent(int howMany) {
  x = Wire.read();
}

static void requestEvent() {
  Wire.write(y++);
}

void setup() {
  Serial.begin(115200);
  Wire.begin(11);                // join i2c bus as slave
  Wire.onRequest(requestEvent);  // register event
  Wire.onReceive(receiveEvent);  // register event
  lastms = millis();
}

void loop() {
  if (millis() > lastms + 1000) {
    Serial.print("-");
    lastms = millis();

  }
  if (x != 0) {
    Serial.printf("U:0x %x\n", x);
    x = 0;
  }
}
```
</details>

Note that I've clean up useless thing and optimize the slave. When you get the request you have to send it immediately else you missed the target.

**Comment by AndKe** [comment]
@fpistm I believe that the commented `Wire.onRequest(requestEvent)` part was just another desperate test.
The main issue is that when acting as slave, (recieving traffic) the device just kept holding SCL low "forever".  

I uploaded the "Slave based on your sketch" that you posted,
On the bus, I have a ArduPilot I2C as master.  it is requesting data: 
![image](https://github.com/user-attachments/assets/6326f986-ec77-4ade-aa0a-9bd58d757a08)

Please note how the SCL is suddenly kept low indefinitely.

Then another capture: the F103 is running fine for several selconds, and then I connect the I2C bus (where the master is talking to it.)
![image](https://github.com/user-attachments/assets/d5506e79-29ba-4d47-853f-0fb096ff40e7)

Please note that in both cases: the F103 is the one that is stuck keeping the SCL low - intil reset - disconnecting master leaves a fully functional I2C master.

Please see the Saleae logic capture - [Saleae capture.zip](https://github.com/user-attachments/files/18169806/Saleae.capture.zip)
You can open the files using https://www.saleae.com/pages/downloads  the tool is multiplatform and free.



When SCL is stuck low like that, the "-" is no longer printed. (most likely the device is crashed/hung/looped at that point)

**Comment by fpistm** [comment]
Please test sketches provided.
I can't reproduce.

And what is your full setup. You talk about Ardupilot but it seems to be a dedicated software. 

What does the Ardupilot software does? Are you sure I2C @ used by the master to request data is the correct one?

**Comment by AndKe** [comment]
@fpistm  
Ardupilot project: https://github.com/ArduPilot/ardupilot
It uses I2C quite much for all kinds of internal sensors, but usually has an external I2C bus for less-critical devices.
To be precise: this is the part I am trying to communicate with: https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp

Yes, I've been emulating that BMS for several years using Atmega 328 and Atmega32u4  with Arduino libraries. 
This whole issue started as I discovered that compiling my code for STM32F103 
(I have used STM32 for many other projects, as I2C master,  and it always worked fine)

That's why we are now trying to do basic stuff like trying to see a write requests as a slave. 

I loaded master sketch on another STM32F103 - but it does not generate I2C traffic. (both lines remain high)

Here is my slave firmware:
[I2C_test_slave.ino.zip](https://github.com/user-attachments/files/18170866/I2C_test_slave.ino.zip)
And my master firmware:
[I2C-test-master.ino.zip](https://github.com/user-attachments/files/18170876/I2C-test-master.ino.zip)

Maybe there's something with the toolchain/build environment?
The fact is that I first had trouble with this in VSCode + PlatformIO , then tried Arduino IDE (now 2.3.4)  - and I see no difference. That's why I thought this is related to the wire library.

**Comment by fpistm** [comment]
With pullup on each I2C lines?

**Comment by AndKe** [comment]
yes, 10k to 5v for both, of course.

**Comment by fpistm** [comment]
Usually 4.7k and stm32 is 3.3v even if gpio are 5v tolerant.

**Comment by AndKe** [comment]
I know, but this is not the issue here, I do not see some odd levels when looking at the signals with an oscilloscope, crisp and clear  - (when there is something to see)  my choice is simply because some of the autopilots use 5v buffers for I/O and then 10k is standard.
Did you find some hardware to install Ardupilot on?   - or run my binary ?  / does your master-build cause any I2C traffic?

**Comment by fpistm** [comment]
I've tested your slave binary and it works on a BluePillF103C8 (with a genuine STM32F103C8) and it works.

> does your master-build cause any I2C traffic?

As stated it works...
My setup is simple 2 boards with I2C lines wired (with 4.7K PU on each I2C lines), GND interconnected and that's all.

Request from the master:
![image](https://github.com/user-attachments/assets/60c6a082-228d-4d16-99f0-355c66ce2666)
then 10 ms later sent from the Master:
![image](https://github.com/user-attachments/assets/09aa60f8-4b2f-4ebb-ac5b-c62b7bdfe304)

> Did you find some hardware to install Ardupilot on?

No, I have no such hardware and will not go further while the basic example not working on your side.
Do you used genuine STM32? Please use 4.7K or less but no 10K....

**Comment by fpistm** [comment]
Finally I've tested with an Ardunio UNO as master (5V) and BluePill with your slave binary and it works.

Edit: I convert this as a discussion

**Comment by AndKe** [comment]
I assume your master traffic is captured on pins PB6/PB7 

This leaves us with two options: 
1: This issue is somehow limited to actual ArduPilot I2C frequency/timing .. but that would be very odd. (and why does it the master fail to not work for me?)

2: This issue started as troubleshooting of devices with LCSC C8734  , assembled by JLCPCB (assembled about 6 weeks ago).  - then got confirmed/reproduced on cheap bluepill's on breadboard (those were bought maybe a year ago)
This option would indicate that both of the uC are the same fake/defective series?


Q1: What do we know about the chance of LCSC being infected by bad/fake stock?  and the chance for the Bluepills to have the same issue? 

Q2: would it help if I posted microscope photos of the LCSC assembled parts?  

Q3: are there some internal serial, some code that can be run to verify the authenticity of the STM32F103  or reveal some known copies ?

I have no other boards with big footprint like this, will try to compile and run the code/software on STM32F405 , An older board with STM32F103C8  - and an STM32F103CB

**Comment by AndKe** [comment]
@fpistm It must have been late yesterday and I must have cocked up something - sorry.
 (maybe I flashed  slave onto both)
Corrected situation is: your master and slave sketch works just fine on my bluepills.
The slave works fine with the example master code above. 
The slave crashes/stops upon receiving request from the actual I2C master:

![image](https://github.com/user-attachments/assets/f27cc291-5df8-4491-b84a-053e695a6e72)

So this *is* some issue, we should not have been able to crash /loop a uC this way.

In this case: we see master request 0x29   - (I think it would read 14bytes.)  the program does not run long enough to even print the request, it hangs before that.

**Reply by fpistm** [reply]
> The slave crashes/stops upon receiving request from the actual I2C master:

You talk about Ardupilot code as master, right?
You could try to disable the STM32 I2C strechmode using:
`Wire.begin(11, false, true);`
What is  the freq used by Ardupilot for I2C ? 100k? 400K ? ...
You can try also:
`Wire.setClock(400000);`

Unfortunately, I have no more clue.

**Comment by AndKe** [comment]
Essential progress:

Please see the following throw-together test:
If you run the code below as master, the slave example you posted previously, that we both had running fine, will crash.
(Your master-test works slave-test, but if the slave-test is connected to the master below, it will crash.)

```c++
#include <Wire.h>

#define BATTMONITOR_SMBUS_ADDRESS 0x0B
#define BATTMONITOR_SMBUS_SOLO_CELL_VOLTAGE_EXT 0x29 

void setup() {
  Serial.begin(9600);
  Wire.begin();
}

void loop() {
  uint8_t buff[12];
  if (read_block(BATTMONITOR_SMBUS_SOLO_CELL_VOLTAGE_EXT, buff, 12)) {
    for (int i = 0; i < 12; i++) {
      Serial.print("0x");
      Serial.print(buff[i], HEX);
      Serial.print(" ");
    }
    Serial.println();
  } else {
    Serial.println("Failed to read data");
  }
  
  delay(1000); 

bool read_block(uint8_t command, uint8_t* data, uint8_t length) {
  Wire.beginTransmission(BATTMONITOR_SMBUS_ADDRESS);
  Wire.write(command);
  if (Wire.endTransmission() != 0) {
    return false;  // End transmission and check for a successful transmission
  }
  
  Wire.requestFrom(BATTMONITOR_SMBUS_ADDRESS, length);
  if (Wire.available() != length) {
    return false;  // Check if the requested amount of data is available
  }
  
  for (uint8_t i = 0; i < length; i++) {
    data[i] = Wire.read();
  }
  
  return true;
}
```
Please confirm that you can reproduce it.

**Reply by fpistm** [reply]
No time today to try to reproduce, anyway this check is not correct:
`  if (Wire.available() != length) {`
This simply told data are available but it is possible not all data you expect are already there.

```c++
  // Maybe you can uncomment this line to ensure first data available
  // while(!Wire.available());
  for (uint8_t i = 0; i < length; i++) {
    if (Wire.available()) {
      data[i] = Wire.read();
    } else {
      return false;
    }
  }
```

**Reply by fpistm** [reply]
Hi @AndKe 
If you used this master example (note it misses a `}` :wink: ) with my slave example this is normal it does not work.
The (my) slave sketch is a mirror of the (my) master which request 1 byte but your sketch requires 12 bytes but my slave send only one byte on master request.

In general a real I2C slave device will get in its internal registers what it have to do and how many data it have to send.
The length give to the   `Wire.`requestFrom`(BATTMONITOR_SMBUS_ADDRESS, length);` will send the length to be written by default in the register 0 of the slave device and it is managed by the hardware itself.

**Reply by AndKe** [reply]
Hello again. Have you seen this https://github.com/orgs/stm32duino/discussions/2608#discussioncomment-11610298
It seems to me like you replayed to something from the past.

**Reply by fpistm** [reply]
> Hello again. Have you seen this https://github.com/orgs/stm32duino/discussions/2608#discussioncomment-11610298 It seems to me like you replayed to something from the past.

Yes, I've only try to answer to all your comment and it is still valid.

**Comment by AndKe** [comment]
@fpistm   
Ardupilot I2C defaults to 100kHz.
Disabling stretch mode seems to help !!  - I have been unable to reproduce the issue since.  So far this is a HUGE success!  I have to spend some family time now, will test more later - but this looks good. 
Now that you can reproduce it (with the example above) - maybe you can figure out what was happening?

**Reply by fpistm** [reply]
> Now that you can reproduce it (with the example above) - maybe you can figure out what was happening?

Never told I can reproduce it. Only try to guess what you do, use and how... which is not easy :wink:

**Comment by AndKe** [comment]
This is the ultimate proof of ..some bug/unintentional issue:

You made things not crash/loop by suggesting:
`You could try to disable the STM32 I2C strechmode using:  Wire.begin(11, false, true);`

That looked as if solved crashing, but then things did not work as intended, I spent some time trying to understand why everything seemed to fall apart.

I have made a good package that tests the communication as if this was the actual device:
master sends some request (one byte)  - and requests ten bytes in return-
slave prints the request returns 10 bytes, the master prints the received bytes.

Please load the following two programs: 

Master:
```c++
//MASTER

#include <Wire.h>

#define SLAVE_ADDRESS 0x0B
int i = 1;

void setup() {
  Wire.begin();  // Join I2C bus as master
  Serial.begin(9600);
  pinMode(PC13, OUTPUT);
}

void loop() {
  Wire.beginTransmission(SLAVE_ADDRESS);
  Wire.write(i++);
  Wire.endTransmission();
  
  // Request 10 bytes from the slave
  Wire.requestFrom(SLAVE_ADDRESS, 10);
  
  // Wait for the data and print it
  Serial.print("Received: ");
  while (Wire.available()) {
    uint8_t byteReceived = Wire.read();
    Serial.print(byteReceived);
    Serial.print(" ");
  }
  Serial.println();
  
  delay(500);  // Wait 500 milliseconds before repeating
  digitalWrite(PC13, !digitalRead(PC13));
}

```
 

Slave:

```c++
//SLAVE

#include <Wire.h>

#define RESPONSE_SIZE 10
volatile uint8_t x = 0;
uint8_t counter = 0;

long last = millis();

void setup() {
  Wire.begin(0x0B); //, false, true);    << this prevents crashing with real devices, but comes with a terrible effect
  Wire.onReceive(receiveEvent);
  Wire.onRequest(requestEvent);
  Serial.begin(115200);
}

void loop() {
  if (millis() > last + 500) {
    Serial.println(x);
    last = millis();
  }
}

// This function is called when the master sends data
void receiveEvent(int howMany) {
  x = Wire.read();
}

// This function is called when the master requests data
void requestEvent() {
  Serial.println("req");
  uint8_t response[RESPONSE_SIZE];
  for (int i = 0; i < RESPONSE_SIZE; i++) {
    response[i] = counter;
    counter++;
    if (counter > 254) {
      counter = 0;
    }
  }
  Wire.write(response, RESPONSE_SIZE);  // Send response
}


```

Please notice the crazy things that are going on once the clock stretching is disabled.
In short: not disabling clock stretching makes it crash - disabling it ...corrupts the data in a crazy way. 
So clock stretching is essential for even this low-complexity computation of return data. 

I made a video for you: https://drive.google.com/file/d/1vYfU9P-NxrzM06YYjbLcQMxBt_xuhyXR/view?usp=sharing

Thank you for looking into this - I hope this helps you understand what is going on here.


-----------------------------------------------------------
Additional data with the real I2C master  and the same sleve test code, but only 4-byte response (also at 100kHz):

With clock stretching disabled: 
The slave was supposed to return 4 bytes, you see request for data packets 0x28 and 0x29 - but not th four bytes being sent.
![Screenshot From 2024-12-18 23-11-59](https://github.com/user-attachments/assets/1a14c9c0-9820-419b-92f2-029837600d01)

With clock stretching not disabled:
I connected the master to the STM32F103 , with analyzer probes already attached to it.
You can see request for 0x28 - with no good response
then request for 0x29 - with another failed response
then another request or 0x29 with what seems to be a good response from slave, before the slave failed and kept SCL low. 
it manages to print three "req" , but none "x"

![image](https://github.com/user-attachments/assets/c6d7896b-2b94-4758-81aa-9f1f6dbf2970)

**Reply by fpistm** [reply]
It works as expected on my side after updated your sketch
Hereafter your sketches with my comments and some advises/optimizations:
```C++
// MASTER

#include <Wire.h>

#define SLAVE_ADDRESS 0x0B
int i = 1;

void setup() {
  Wire.begin();  // Join I2C bus as master
  Serial.begin(9600);
  pinMode(LED_BUILTIN, OUTPUT);
}

void loop() {
  Wire.beginTransmission(SLAVE_ADDRESS);
  Wire.write(i++);
  Wire.endTransmission();

  // Request 10 bytes from the slave
  Wire.requestFrom(SLAVE_ADDRESS, 10);

  // Wait for the data and print it
  if (Wire.available()) {
    Serial.print("Received: ");

    while (Wire.available()) {
      uint8_t byteReceived = Wire.read();
      Serial.print(byteReceived);
      Serial.print(" ");
    }
    Serial.println();
  }

  delay(500);  // Wait 500 milliseconds before repeating
  digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
}

```

```C++
// Slave
#include <Wire.h>

#define RESPONSE_SIZE 10
const uint8_t response[RESPONSE_SIZE] {0,1,2,3,4,5,6,7,8,9};
volatile uint8_t x = 0;
static uint8_t x_prev = 0;
uint8_t counter = 0;

// long last = millis();

void setup() {
  Wire.begin(0x0B);  //, false, true);    << this prevents crashing with real devices, but comes with a terrible effect
  Wire.onReceive(receiveEvent);
  Wire.onRequest(requestEvent);
  Serial.begin(115200);
}

void loop() {
  // Print only if x changed
  // if (millis() > last + 500) {
  if (x != x_prev) {
    Serial.println(x);
    //last = millis();
    x_prev = x;
  }
}

// This function is called when the master sends data
void receiveEvent(int howMany) {
  (void)howMany;
  x = Wire.read();
}

// This function is called when the master requests data
void requestEvent() {
  // Here only performed data processing.
  // I do not advise to use Serial in a cb which is time consuming without any interest
  // preferably use debug in that case or a gpio toggle.
  // Serial.println("req");

  // Why always init with same values at each request.
  // Anyway I guess the compiler optimizes as a constant array.

  // uint8_t response[RESPONSE_SIZE];
  // for (int i = 0; i < RESPONSE_SIZE; i++) {
  //   response[i] = counter;
  //   counter++;
  //   // Guess this test is in case you change RESPONSE_SIZE
  //   if (counter > 254) {
  //     counter = 0;
  //   }
  // }
  Wire.write(response, RESPONSE_SIZE);  // Send response
}

```

**Reply by fpistm** [reply]
To go further hereafter examples with different request size from master and the slave answering based on the value received.

Master send the num of the response it will request after.

```c++
// MASTER
#include <Wire.h>

#define SLAVE_ADDRESS 0x0B
static int i = 1;

void setup() {
  Wire.begin();  // Join I2C bus as master
  Serial.begin(9600);
  pinMode(LED_BUILTIN, OUTPUT);
}

void loop() {
  Wire.beginTransmission(SLAVE_ADDRESS);
  Wire.write(i);
  Wire.endTransmission();

  // Request 10 bytes from the slave
  Wire.requestFrom(SLAVE_ADDRESS, i);

  // Wait for the data and print it
  if (Wire.available()) {
    Serial.printf("Received for request %i: ", i++);

    while (Wire.available()) {
      uint8_t byteReceived = Wire.read();
      Serial.printf("%u ", byteReceived);
    }
    Serial.println();
  }

  delay(500);  // Wait 500 milliseconds before repeating
  digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
  if (i == 5) {
    i = 1;
  }
}
```

```c++
#include <Wire.h>

#define RESPONSE_SIZE1 1
#define RESPONSE_SIZE2 2
#define RESPONSE_SIZE3 3
#define RESPONSE_SIZE4 4
const uint8_t response1[RESPONSE_SIZE1]{ 1 };
const uint8_t response2[RESPONSE_SIZE2]{ 2, 3 };
const uint8_t response3[RESPONSE_SIZE3]{ 4, 5, 6 };
const uint8_t response4[RESPONSE_SIZE4]{ 7, 8, 9, 10};
volatile uint8_t x = 0;

void setup() {
  Wire.begin(0x0B);  //, false, true);    << this prevents crashing with real devices, but comes with a terrible effect
  Wire.onReceive(receiveEvent);
  Wire.onRequest(requestEvent);
  Serial.begin(115200);
}

void loop() {
}

// This function is called when the master sends data
void receiveEvent(int howMany) {
  (void)howMany;
  x = Wire.read();
}

// This function is called when the master requests data
void requestEvent() {
  switch (x) {
    default:
    case 1:
      Wire.write(response1, RESPONSE_SIZE1);  // Send response
      break;
    case 2:
      Wire.write(response2, RESPONSE_SIZE2);  // Send response
      break;
    case 3:
      Wire.write(response3, RESPONSE_SIZE3);  // Send response
      break;
    case 4:
      Wire.write(response4, RESPONSE_SIZE4);  // Send response
      break;
  }
}
```
![image](https://github.com/user-attachments/assets/8856d7c5-9c25-441f-bc01-c9f4c16f3a6d)

**Comment by AndKe** [comment]
@fpistm hmm, I'd say you cheated.  :) 
The whole idea/need for clock stretching is basically worked around.
In my application to do it "this way"  I'd need to have pre-processed different kinds of replies for all kinds of challenges. which will at least occupy more RAM. 
I prefer to assemble the response bytes and calculate check sums on demand (on requestEvent).  (the to be sent data is sampled at another rate, and is filtered, and first now is split into correct bytes, checksums calculated etc)

I am also afraid that if the BMS is connected to the running master  (think hotswap/groundpower) then it will crash before all the data and calculations can be done (or just be forced to send invalid/empty data while the program is still in requestEvent)  

As for using print in interrupts, I did it here because testing proved I could get away with it in this case :)  

But I would like to be able to do proper data calculation before sending.  (and be able to rely on clock-stretching until done)  - in another application I use, where the data needs to be as fresh as possible, and cannot be sampled at a high rate, this is kind of essential.  (to be able to receive a request, compute for a while let's say  (0.1-1ms) , then respond.)

Also: it will require twice the size of each buffer/each reply to be able to update one buffer/reply, and have another be valid at any given moment. (otherwise I risk sending a buffer that is in the process of being updated)

Basically: this should work with some ms delay:

void receiveEvent(int howMany) {
  (void)howMany;
  x = Wire.read();
/// here we need some processing delay. 
}

**Reply by AndKe** [reply]
@fpistm  please see my reply above. I verified your latest example.

Adding this one 100uS delay messes things up.

void receiveEvent(int howMany) {
  (void)howMany;
  x = Wire.read();
  delayMicroseconds(100);
}

In essence: we have no working clock stretching. This would require having everything pre-calculated and doubled up (so that a buffer in process of being updated was not transmitted.)  Also:  this kills any idea of low-data-age transmissions.  as there will be jitter between when data is acquired and packed, and actually requested.

Please make the clock-stretching work the way it normally does.  So that the master can properly wait for the data to be prepared.

**Reply by fpistm** [reply]
@AndKe 
Adding a delay or other data processing is not a good idea, this function is called inside ISR and goal is only to alert user program data have been received. I don't think Clock stretching is useful, here.
You should simply set a boolean to told that data has been received then process in your application outside the ISR.
Anyway, STM32F1 have some issue related to I2C and maybe we fall in those case it.

I will try to dig more on this next year. Any contribution on this subject are welcome as it is a community project. As stated in the [stm32duino home](https://github.com/stm32duino#welcome-to-the-stm32duino-open-source-community
)
> [!NOTE]  
> STM32duino GitHub organization is an open source community, it is not part of the official software ecosystem supported by ST.
> Nevertheless ST contributes to this community.

**Reply by AndKe** [reply]
@fpistm Of course, it is simply "better" to do it inside the ISR as the data is as fresh as possible, and, there is no chance of transmitting half-done job. (some of the bytes updated, but not all)
The clock stretching is absolutely essential for many I2C devices that cannot or should not deliver data at random, but sample/prepare them on request.

FWIW: This feature does work perfectly with other MCU's  - like mentioned before, I had it running using Wire library on Atmega328 and 16U4 just fine until I needed to modernise and throw a better/more modern uC at the job. 
I wish you a great holiday and hope that this issue will be acknowledged as such, and fixed sometime in 2025 :) 
Merry Christmas

**Reply by fpistm** [reply]
Thanks, you too.
Receive and request are differents.
When you process inside isr other interrupts processing are not executed except ones with higher priorities.

So don't understand your point with slave receive.when master send data it is received by the slave...

**Reply by AndKe** [reply]
Well, I would not claim to have the answer about "best practice" about this, but having clock stretching on  a slave is a very important thing for a slave device that actually does a processing/computing job, or needs to produce updated data on demand, and does not have everything ready at any given time.  This is why this is a basic part of the I2C bus and it's derivatives. (like PMBUS or DDC)  - I'd almost say that without clock stretching we are not really I2C compliant.  (which I believe is the case).

**Reply by fpistm** [reply]
Yes but in the case of receive it simply slow down the clock if needed. When the receiveEvent is called data already received by the slave and available in the wire buffer. So no issue with stretch here. 😉

**Comment by AndKe** [comment]
@fpistm  Hope you had a great holiday.
I did see, but not fully understand your answer here: 
"Yes but in the case of receive it simply slow down the clock if needed. When the receiveEvent is called data already received by the slave and available in the wire buffer. So no issue with stretch here. 😉"

Now that we have narrowed it down to being perfectly reproducible, we established two facts: 

1: Not even  even 0.2us of processing time is allowed from an interrupt triggers requestEvent()   -  normally, clock stretching should hold hold the SCL low for a while, not indefinitely.     

2: there is no way not to respond to a write request, is Wire.onRequest is jumper to by interrupt, but does not have anything to transmit/is not ready/  - it all fails as described above.

This is not really I2C compliant, and can only serve prepossessed / "old" data.

**Comment by shtirlic** [comment]
Interesting thread, had the same issue, see the pic.
Master: rp2350 on Nuttx (m33 core)
Slave: stm32f103r8t6
Device: https://www.clockworkpi.com/product-page/picocalc
speed:  Just 10kHz
![image](https://github.com/user-attachments/assets/4aed97a1-3a78-4d7c-b30b-b01296b690a3)
Disabling strethcmode helps, but brokes other i2c masters on the same rp2350 but different implementations.
Tried to increase i2c timeout on pico2 side(master), tried different tLow and tHigh, tried custom sda tx hold count, nothing helps yet.  

PS
Also thinking about something interrupting the stm32 during requestEvent and it stalls somehow stm32 i2c bus
